homeassistant/components/mqtt/entity.py

```diff
@@ -703,12 +705,63 @@ async def async_discovery_update(
     ) -> None:
         """Handle discovery update."""
         discovery_hash = get_discovery_hash(self._discovery_data)
+        # Start discovery migration or rollback if migrate_discovery flag is set
+        # and the discovery topic is valid and not yet migrating
+        if (
+            discovery_payload.migrate_discovery
+            and self._migrate_discovery is None
+            and self._discovery_data[ATTR_DISCOVERY_TOPIC]
+            == discovery_payload.discovery_data[ATTR_DISCOVERY_TOPIC]
+        ):
+            self._migrate_discovery = self._discovery_data[ATTR_DISCOVERY_TOPIC]
+            discovery_hash = self._discovery_data[ATTR_DISCOVERY_HASH]
+            if discovery_payload.device_discovery:
+                _LOGGER.info(
+                    "Rollback started for device based discovery migration "
+                    "on topic %s. Publish a component based "
+                    "discovery update for component %s to roll back",
+                    self._migrate_discovery,
+                    discovery_hash,
+                )
+            else:
+                _LOGGER.info(
+                    "Migration to device based discovery started "
+                    "on topic %s. Publish a device based discovery update "
+                    "including component %s",
+                    self._migrate_discovery,
+                    discovery_hash,
+                )
+            # Cleanup platform resources
+            await self.async_tear_down()
+            # Unregister and clean discovery
+            stop_discovery_updates(
+                self.hass, self._discovery_data, self._remove_discovery_updated
+            )
+            send_discovery_done(self.hass, self._discovery_data)
+            return
+
         _LOGGER.debug(
             "Got update for %s with hash: %s '%s'",
             self.log_name,
             discovery_hash,
             discovery_payload,
         )
+        new_discovery_topic = discovery_payload.discovery_data[ATTR_DISCOVERY_TOPIC]
+
+        # Abort early if there is no normal update
```

Comment:
I'm not sure what a "normal update" means, can you try to clarify?

Reply:
done